### PR TITLE
Dat 578 prepaid for free users

### PR DESCRIPTION
**Pull request by msimionati**
new checkout support to buy credits for free users

### CODE REVIEW

File: Doppler.BillingUser.Test/PostAgreementTest.cs [hunk 1]
```diff
@@ -32,54 +49,683 @@ public async Task Create_agreement_return_ForbidResult_when_total_amount_validat
...
+                PlanId = 2
+            };
+
+            var creditCard = new CreditCard()
+            {
+                CardType = CardTypeEnum.Visa,
+                ExpirationMonth = 12,
+                ExpirationYear = 23,
+                HolderName = "kBvAJf5f3AIp8+MEVYVTGA==",
+                Number = "Oe9VdYnmPsZGPKnLEogk1hbP7NH3YfZnqxLrUJxnGgc=",
```
Comment (andresmoschini) [inline Doppler.BillingUser.Test/PostAgreementTest.cs:100]: No es un número real, no?

Comment (msimionati) [reply Doppler.BillingUser.Test/PostAgreementTest.cs:100]: Es el numero de prueba de visa

File: Doppler.BillingUser/Model/AgreementInformation.cs [hunk 2]
```diff
@@ -4,7 +4,7 @@ public class AgreementInformation
     {
         public int PlanId { get; set; }
         public int DiscountId { get; set; }
-        public decimal Total { get; set; }
+        public decimal? Total { get; set; }
```
Comment (nreal) [inline Doppler.BillingUser/Model/AgreementInformation.cs:7]: Why total could be null?

File: Doppler.BillingUser.Test/PostAgreementTest.cs [hunk 3]
```diff
@@ -32,54 +49,699 @@ public async Task Create_agreement_return_ForbidResult_when_total_amount_validat
...
+                    services.AddSingleton(userRepositoryMock.Object);
+                    services.AddSingleton(paymentGatewayMock.Object);
+                    services.AddSingleton(billingRepositoryMock.Object);
                 });
             }).CreateClient(new WebApplicationFactoryClientOptions());
             client.DefaultRequestHeaders.Add("Authorization", $"Bearer {TOKEN_ACCOUNT123_TEST1_AT_TEST_DOT_COM_EXPIRE20330518}");
 
             // Act
-            var response = await client.PostAsJsonAsync("accounts/test1@test.com/agreements", agreement);
+            var response = await client.PostAsync($"accounts/{accountName}/agreements", JsonContent.Create(agreement));
```
[the remark this answers is not in the document]
Comment (msimionati) [reply Doppler.BillingUser.Test/PostAgreementTest.cs:144]: Do you meen using the PostAsJsonAsync?

File: Doppler.BillingUser/Controllers/BillingController.cs [hunk 4]
```diff
@@ -119,13 +131,77 @@ public async Task<IActionResult> UpdateCurrentPaymentMethod(string accountname,
...
+                var authorizationNumber = await _paymentGateway.CreateCreditCardPayment(agreementInformation.Total.GetValueOrDefault(), encryptedCreditCard, user.IdUser);
+                await _billingRepository.CreateAccountingEntriesAsync(agreementInformation, encryptedCreditCard, user.IdUser, authorizationNumber);
+            }
+
+            var billingCreditId = await _billingRepository.CreateBillingCreditAsync(agreementInformation, user, newPlan);
+
+            user.IdCurrentBillingCredit = billingCreditId;
+            await _userRepository.UpdateUserBillingCredit(user);
+
+            if (agreementInformation.Total.HasValue)
```
Comment (andresmoschini) [inline Doppler.BillingUser/Controllers/BillingController.cs:196]: Me parece que lo que querés comprobar no es que tenga valor, sino que el valor sea mayor que cero.

```suggestion
            if (agreementInformation.Total.GetValueOrDefault() > 0)
```

Antes de arreglarlo, estaría bueno hacer tests que cubran este escenario:

* Validar que cuando el valor sea mayor que cero se haga todo (creo que está hecho)
* Validar que cuando el valor sea cero determinadas cosas no se hagan

File: Doppler.BillingUser/Model/AgreementInformation.cs (hunk 2 above)
Comment (andresmoschini) [reply Doppler.BillingUser/Model/AgreementInformation.cs:7]: I am responsible for this.

If you make this field not nullable, when the user forgot to send this value in the JSON, .NET assumes zero.

If we want to validate if the user send this value or not, the easy way is to make it nullable.

File: Doppler.BillingUser/Infrastructure/BillingRepository.cs [hunk 5]
```diff
@@ -378,5 +392,375 @@ LEFT JOIN
...
+                ApprovedDate = now,
+                Approved = true,
+                Payed = true,
+                IdUserTypePlan = agreementInformation.PlanId,
+                PlanFee = (double?)agreementInformation.Total,
+                CreditsQty = newUserTypePlan.EmailQty ?? null,
+                ExtraEmailFee = newUserTypePlan.ExtraEmailCost ?? null
+            };
+
+            var utcNow = DateTime.UtcNow;
```
Comment (andresmoschini) [inline Doppler.BillingUser/Infrastructure/BillingRepository.cs:537]: Ya definiste `now`, creo que lo mejor es seguir usando el mismo valor.

```suggestion
            var now = DateTime.UtcNow;
            buyCreditAgreement.BillingCredit = new BillingCreditModel()
            {
                Date = now,
                PaymentDate = now,
                ApprovedDate = now,
                Approved = true,
                Payed = true,
                IdUserTypePlan = agreementInformation.PlanId,
                PlanFee = (double?)agreementInformation.Total,
                CreditsQty = newUserTypePlan.EmailQty ?? null,
                ExtraEmailFee = newUserTypePlan.ExtraEmailCost ?? null
            };
```

File: Doppler.BillingUser.Test/PostAgreementTest.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (andresmoschini) [reply Doppler.BillingUser.Test/PostAgreementTest.cs:144]: Yes, it looks simpler, right?

**Comment by andresmoschini** [comment]
Nice work! 

I detected a scenario that is not well covered by the tests (https://github.com/FromDoppler/doppler-billing-user/pull/56#discussion_r765851865), please review if there are more scenarios like this one.

File: Doppler.BillingUser/Infrastructure/BillingRepository.cs (hunk 5 above)
Comment (msimionati) [reply Doppler.BillingUser/Infrastructure/BillingRepository.cs:537]: Tenes razón, se me paso..., quedo así porque antes eran dos métodos que terminaron en uno solo. 
Lo saco.

File: Doppler.BillingUser/Controllers/BillingController.cs (hunk 4 above)
Comment (msimionati) [reply Doppler.BillingUser/Controllers/BillingController.cs:196]: Entiendo, voy a implementar unos test que cubran el escenario.

File: Doppler.BillingUser/Model/AgreementInformation.cs (hunk 2 above)
Comment (msimionati) [reply Doppler.BillingUser/Model/AgreementInformation.cs:7]: @andresmoschini you beat me, I was looking for a way to explain it.
"This was a way to handle the situation of the zero default value of the int type, as the total can be zero, the idea is that the "zero" value is set explicitly from the client and that it is not a situation where a default value drives the flow of the agreement creation."

File: Doppler.BillingUser/Infrastructure/BillingRepository.cs (hunk 5 above)
Comment (msimionati) [reply Doppler.BillingUser/Infrastructure/BillingRepository.cs]: Listo, se cambio.

File: Doppler.BillingUser/Controllers/BillingController.cs (hunk 4 above)
Comment (msimionati) [reply Doppler.BillingUser/Controllers/BillingController.cs]: Listo, agregue una prueba, que permite el total cero y valida que no se creen los registros asociados a un pago con tarjeta de crédito.

File: Doppler.BillingUser.Test/PostAgreementTest.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (msimionati) [reply Doppler.BillingUser.Test/PostAgreementTest.cs]: Done, modified

File: Doppler.BillingUser/Controllers/BillingController.cs (hunk 4 above)
Comment (andresmoschini) [reply Doppler.BillingUser/Controllers/BillingController.cs]: ❤️

**Comment by andresmoschini** [comment]
:shipit:

**Pull request merged**